data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,670 @@
+            Self::overflow()
+        }
+    }
+
+    /// There is little that can be done if an allocation fails
+    /// due to incorrect layout or memory exhaustion.
+    #[allow(clippy::panic)]
+    #[inline]
+    const fn overflow() -> ! {
+        panic!("overflow");
+    }
+}
+
+impl AsRef<str> for BoxedString {
+    #[allow(unsafe_code)]
+    #[inline]
+    fn as_ref(&self) -> &str {
+        // SAFETY: created from valid utf-8
+        unsafe { from_utf8_unchecked(self.as_bytes()) }
+    }
+}
+
+impl Deref for BoxedString {
+    type Target = str;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl Clone for BoxedString {
+    /// Properly clone `BoxedString` into new allocation.
+    #[allow(unsafe_code)]
+    fn clone(&self) -> Self {
+        let len = self.len;
+        let layout = Self::layout(len);
+        // SAFETY: pointer is checked to be non-null.
+        let ptr = unsafe { alloc::alloc::alloc(layout) };
+        if ptr.is_null() {
+            Self::overflow();
+        }
+        // SAFETY: all conditions are met.
+        unsafe { ptr.copy_from_nonoverlapping(self.ptr, self.len) };
+        Self { ptr, len }
+    }
+}
+
+impl From<&str> for BoxedString {
+    #[allow(unsafe_code)]
+    #[inline]
+    fn from(value: &str) -> Self {
+        let len = value.len();
+        let boxed = value.as_bytes().to_vec().into_boxed_slice();
+        let ptr = Box::into_raw(boxed).cast::<u8>();
+        Self { ptr, len }
+    }
+}
+
+impl From<String> for BoxedString {
+    #[inline]
+    fn from(value: String) -> Self {
+        let len = value.len();
+        let boxed = value.into_boxed_str();
+        let ptr = Box::into_raw(boxed).cast::<u8>();
+        Self { ptr, len }
+    }
+}
+
+impl Drop for BoxedString {
+    #[allow(unsafe_code)]
+    fn drop(&mut self) {
+        // SAFETY:
+        // - using global allocator
+        // - `BoxedString`s have the same layout varied only by it's length
+        unsafe {
+            alloc::alloc::dealloc(self.ptr as *mut u8, Self::layout(self.len));
+        }
+    }
+}
+
+/// `BoxedString` is `Send` because the data they
+/// reference is unaliased. Aliasing invariant is enforced by
+/// creation of `BoxedString`.
+#[allow(unsafe_code)]
+unsafe impl Send for BoxedString {}
+
+/// `BoxedString` is `Send` because the data they
+/// reference is unaliased. Aliasing invariant is enforced by
+/// creation of `BoxedString`.
+#[allow(unsafe_code)]
+unsafe impl Sync for BoxedString {}
+
+#[derive(Clone, Copy)]
+#[repr(C)]
+struct InlinedString {
+    #[cfg(target_endian = "little")]
+    payload: [u8; MAX_INLINED_STRING_LEN],
+    /// MSB is always 1 to distinguish inlined variant.  
+    len: u8,
+    #[cfg(target_endian = "big")]
+    payload: [u8; MAX_INLINED_STRING_LEN],
+}
+
+impl InlinedString {
+    #[inline]
+    const fn len(self) -> usize {
+        (self.len - 128) as usize
+    }
+
+    #[inline]
+    const fn is_inlined(self) -> bool {
```

Comment:
this should just return true or not exist at all.

The fact that you first cast into an `InlinedString` and then check for the first bit of len is a code smell to me because the underlying bit pattern might not be a valid string. Considering that the string is backed by byte slice all bit patterns might be allowed even though they may not be a valid string. 

There is a [difference](https://internals.rust-lang.org/t/two-kinds-of-invariants-safety-and-validity/8264/23) between unsafety and validity and I'm not sure. Try to find out what's appropriate here. You could cast the `ConstStringData` to a byte array and read the byte instead

Reply:
let's consider this a little bit more before making a decision

Reply:
looking at [this PR](https://github.com/rust-lang/reference/pull/792/files) it may be the case that non-UTF strings don't cause UB

Reply:
I'd tentatively consider this a non-issue as far as I can conclude after researching into the matter